--- texk/web2c/web2c/splitup.c.orig	2016-11-23 13:19:20.000000000 -0500
+++ texk/web2c/web2c/splitup.c	2016-11-23 13:24:40.000000000 -0500
@@ -120,6 +120,9 @@
   if (STREQ (output_name, "mf")) {
     fputs ("#define INIMF\n#define MF\n#define onlyMF\n", out);
     coerce = "mfcoerce.h";
+  } else if (STREQ (output_name, "mp")) {
+    fputs ("#define INIMP\n#define MP\n#define onlyMP\n", out);
+    coerce = "mpcoerce.h";
   } else if (STREQ (output_name, "mflua")) {
     fputs ("#define INIMF\n#define MF\n#define MFLua\n", out);
     coerce = "mfluacoerce.h";
